File: tabulation/cancons.py
# solve the same in js.

def canConstruct(target,words):
    table = [False] * (len(target)+1)
    table[0] = True
    for i in range(len(target)+1):
        if table[i] is True:
            for w in words:
                # print(target)
                print(target[i:len(w)])
                # print(w)
                if target[i:len(w)] == w:
                    # try:
                    table[i+len(w)] = True
                        # print(target[i:len(w)])
                        # print(w)
                    # except:
                    #     pass
    return table
    # [len(target)]

print(canConstruct('abcdef',['ab','abc','abcd','def','cd']))
# print(canConstruct('skateboard',['s','k','t','ate','bo','boar']))
# print(canConstruct('eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeef',['e','ee','eee','eeee','eeeee','eeeeee']))

File: tabulation/cancons.js
const canConstruct = (target,words) => {
    const table = Array(target.length + 1).fill(false);
    table[0] = true;
    for (let i=0; i<=target.length;i++){
        if(table[i] === true){
            for(let w of words){
                if(target.slice(i,i+w.length) === w){
                    table[i+w.length] = true;
                }
            }
        }
    }
    return table[target.length];
}

console.log(canConstruct('abcdef',['ab','abc','abcd','def','cd']))
console.log(canConstruct('skateboard',['s','k','t','ate','bo','boar']))
console.log(canConstruct('eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeef',['e','ee','eee','eeee','eeeee','eeeeee']))
